Parallelize tests

dan-mm opened the pull request:
Not ready yet, making PR to test in VM

dan-mm committed:
initial commit for parallization; introced parallel_id to Runner arguements; updated tests to account for parallel_id in container names

dan-mm committed:
changed conftest.py test cleanup to be per actual test session, not worker session; more cleanup functions editing so parallel tests can run; made test_jobs sequential (only way to ensure they are run on the same worker)

dan-mm committed:
CI jobs to use parallel tests

dan-mm committed:
fix for test_uri_local_dir

dan-mm committed:
add parallel_id to networks; revert uri test fix; try different method for ensuring test_jobs.py are in the same group; updated documentation and pytest invocation command

ArneTR commented:
I do understand it is not ready yet. Two questions still:
- Can you make a draft PR in the future for these? I think it is even more clear to outsiders and us internally what the state of the PR is then. Looks like this: https://github.com/green-coding-berlin/green-metrics-tool/pull/630
- The implementation route is now so that the runner.py is changed. Wasn't the idea to patch the usage_scenario and variables of the runner at runtime? Or does the PR not reflect this discussion we had yet?

dan-mm committed:
fix network check test

dan-mm committed:
Merge branch 'main' into parallelize-tests

dan-mm committed:
- conftest cleanup now applies only after all workers finish
- base test/tmp directory remains, only subfolders are cleaned
- updated run-tests script with new pytest run commands
- added dummy metric providers for tests (same as cpuprocfs provider, but skips base.py check_system as that causes collisions during paralleizations)
- temporarily turn off test_volume_loading::test_volume_loading_subdirectories_subdir2 test as that is failing for unknown reasons
- added serial pytest mark for tests that cannot be parallelized (test_jobs)
- removed now unnecessary jobs table cleanup in test_jobs
- removed duplicate test in test_jobs

dan-mm commented:
> * Can you make a draft PR in the future for these? I think it is even more clear to outsiders and us internally what the state of the PR is then. Looks like this: [Monitor mode #630](https://github.com/green-coding-berlin/green-metrics-tool/pull/630)
Aye, I can make drafts for the future

> * The implementation route is now so that the runner.py is changed. Wasn't the idea to patch the usage_scenario and variables of the runner at runtime? Or does the PR not reflect this discussion we had yet?
I think there was a miscommunication here, I thought we had agreed that the runner.py implementation was to be changed? Patching the usage_scenario will not be enough, as we also need to change the /tmp/green-metrics-tool folder that gets used to be unique.

The discussion regarding patching the usage_scenario was in response to a checking I was doing against the container names from the flow section to match against the parallel_id. You pointed out that patching the object at runtime would make more sense than regex matching, but I thought this was still in context of within the runner.py - which is what I've implemented.

dan-mm committed:
update pytest command for workflow

dan-mm committed:
update workflow test command ;split test action into regular gmt tests and examples directory tests

dan-mm committed:
changed to | tee -a so we can see the output

dan-mm committed:
WIP parallelize rewrite to not edit runner.py, but use parallel_id only in test_functions

dan-mm committed:
remove parallel_id from runner; move custom loader and supporting functions to helper files; test_function's setup_runner now parallizes loaded yml files;

dan-mm committed:
- updated pytest runstring; setup tests to use new test_functions with parallelization correctly; gmt tmp folders now have parallel id pre-pended, not appended; fix issue with yml dumping during parallelization where key order was not preserved; better default parameters for setup_runner; added flags for setup_runner to fine-grain how it is used

dan-mm committed:
smoke tests now use temp directory instead of stress-application directory; - isntead of _ for parallel id; depends_on now properly writes yaml in test parallelization; setup_runner doesn't override uri if its passed in; yml_parsing tests serialized;

dan-mm committed:
parallelize the tmp docker image

dan-mm committed:
serialize failing test

dan-mm committed:
removed leftover unneeded fstring

dan-mm committed:
fix typo in test_volume_loading_subdirectories_root failing test (sanity officially destroyed); update reference to green-coding-solutions in test_jobs

dan-mm committed:
don't run examples directory tests

dan-mm committed:
debug statement

dan-mm committed:
- found tests that were being run without no_build flag when they should have been; setup_runner now properly runs with no_build flag by default; cleaned up calls to setup_runner; build gcb_stress_gmt_run_tmp image before test runs; revert _tmp_image_name parallelization (unneeded);

dan-mm committed:
corrected workflow input check syntax

dan-mm committed:
capitalization

dan-mm committed:
github workflow inputs arent real bools

dan-mm committed:
fix test_jobs (improper cleanup after insert job test); cleanup test directories in /tmp/;

dan-mm committed:
updated tests Readme

dan-mm commented:
@ArneTR - ready for review!

ArneTR commented:
- Can you give me a summary for that commit / design description
- Also: Can you tell me how many tests are parallelized in the end?
  - I see that the test time on Github Actions has gone down from 7 Minutes to 5.30 minutes. Is that what you are seeing on local too? Is that what you expect?

dan-mm commented:
Summary:

- test_functions does the setup_runner differently now: it parses through the usage_scenario and docker compose files that are passed in, and appends a parallel_id (usually a random 12 digit number) to any container names and networks
- after creating a Runner object, it also edits its _tmp_folder and _folder to also be unique (tmp/gmt_tests-{parallel_id}/green-metrics-tool/")

- not all tests were properly using dev_no_build as that wasn't the default in setup_runner (this was the root cause of the docker issue I was seeing). dev_no_build is now the default

- tests that need a uri now more uniformly create (and cleanup) a tmp folder in the tests subdirectory to use as its project folder

- most tests had to be updated to properly check for the parallel_id in their expectations and call the setup_runner correctly

- the base gcb_stress_gmt_tmp_image is now created before any tests are run in conftest.py
- similarly the conftest.py cleanup step is a bit more thorough (getting rid of anything under tests/tmp and clearing db, and done once after all tests are finished)

- I have added a pytest marker (pytest.mark.serial) to mark any tests that need to be run in serial. I've added documentation in the tests/Readme regarding this and running the tests now.

- Custom yaml loader moved from inline definition in runner.py into yml_helper.py file
- similarly, join_paths also moved to utils.py
- I was reusing these functions to the the usage_scenario yaml loading/parsing during setup_runner



As for the reduction, it is what I expect. These are number and times of the kinds of tests:

Parallel: 57 tests; time: 0:01:23
Serial: 15 tests; time: 0:01:01

so of that 5.30, 2.24 minutes test time, 3,06 workflow overhead

From the last main branch workflow, 72 tests; 0:03:41 test time and 3,32 workflow overhead. Which is almost 50% reduction in the test time itself.

It is what I expect as I see a similar percentage reduction in local:

Serialized: 0:07:16 total test time
With parallelization (+ the 15 serial): 0:02:11 + 0:02:09, 4:20 total test time



However.... I have to take back that this PR is ready. When I was running the tests locally to benchmark these times some more I'm seeing some sporadic failures still. It didn't show up when I tested earlier today during pushing, or in the PR tests, but they're there. As best as I can tell right now I think some cleanups are happening early, but I'm not sure yet. At least one more commit is incoming.

dan-mm committed:
removed unneeded dummy cpu util provider; renamed RUN_NAME -> name for readability

dan-mm committed:
Merge branch 'main' into parallelize-tests

ArneTR commented:
@ribalba Please give it a review and validate if the approach is feasible and satisfies the complexity for the expected gains. Then I would like to plan further steps on this

ArneTR commented:
I wil make this PR as a draft to have it as concept in the backlog.

The idea and general implemementation of the parallelization is nice but the PR has some issues:
- Currently it is experiencing unknown race conditions
- many additional refactorings and small changes are added which are unclear if they might be causal for the breaking functionality or just refactorings
- Time gain is not as big as expected due to many synchronous downgrades

ArneTR commented:
Closing this as heavily outdated
